[Merged by Bors] - Prepare proposer

paulhauner opened the pull request:
## Issue Addressed

Resolves #2936

## Proposed Changes

Adds functionality for calling [`validator/prepare_beacon_proposer`](https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Validator/prepareBeaconProposer) in advance.

There is a `BeaconChain::prepare_beacon_proposer` method which, which called, computes the proposer for the next slot. If that proposer has been registered via the `validator/prepare_beacon_proposer` API method, then the `beacon_chain.execution_layer` will be provided the `PayloadAttributes` for us in all future forkchoiceUpdated calls. An artificial forkchoiceUpdated call will be created 4s before each slot, when the head updates and when a validator updates their information.

Additionally, I added strict ordering for calls from the `BeaconChain` to the `ExecutionLayer`. I'm not certain the `ExecutionLayer` will always maintain this ordering, but it's a good start to have consistency from the `BeaconChain`. There are some deadlock opportunities introduced, they are documented in the code.

## Additional Info

- ~~Blocked on #2837~~

paulhauner committed:
Add `proposers` cache to EL

paulhauner committed:
Add proposer prep to beacon chain

paulhauner committed:
Start adding payload prep delay

paulhauner committed:
Repair after rebase

paulhauner committed:
Add fc call to preparation endpoint

paulhauner committed:
Add proposer prep service

paulhauner committed:
Add prepare call to HTTP api

paulhauner committed:
Tidy

paulhauner committed:
Call prepare_proposer after head change

paulhauner committed:
Reduce nesting

paulhauner committed:
Don't start update service if there's no EL

paulhauner committed:
Tidy, add comments

paulhauner committed:
Add forkchoice lock

paulhauner committed:
Upgrade error to dbg

paulhauner committed:
Save prev request in exec layer

paulhauner committed:
Add more tests

paulhauner committed:
Read proposer during preparation

paulhauner committed:
Fix bugs

paulhauner committed:
Fix failing test

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3673,69 +3674,337 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            //
+            // This seems OK. It's not a significant waste of EL<>CL bandwidth or resources, as
+            // far as I know.
+            if let Err(e) = self.prepare_beacon_proposer_blocking() {
```
What's the benefit of running this after `update_execution_engine_forkchoice_blocking`? This seems intentional, but I am only really seeing the potential for duplicate `notify_forkchoice_updated` requests, which like you point out doesn't seem *bad* but I'm not seeing why it might be useful

paulhauner replied:
So `prepare_beacon_proposer_blocking` will only send a fcU (forkchciceUpdated) message to the EL if something has changed or if the timing is right.

As it's currently written, we need to call `prepare_beacon_proposer_blocking` and `notify_forkchoice_updated` to ensure that at least one fcU call goes through to the EE.

We could probably avoid this by breaking out the fcU call from `prepare_beacon_proposer_blocking` it get it to return a "should call fcU bool". I'm not certain it's worth it, but I'll give it some thought.

realbigsean replied:
Calling fcU before calculating the proposer in order to notify the EE as soon as possible seems reason enough to me to keep it as is 👌

realbigsean commented on beacon_node/beacon_chain/src/proposer_prep_service.rs:
```diff
@@ -0,0 +1,65 @@
+    }
+}
+
+/// Loop indefinitely, calling `BeaconChain::prepare_beacon_proposer_async` at an interval.
```
It seems like this service covers the scenario where we have a skip slot right before we propose, and want to make sure we still notify the execution engine of our `PayloadAttributes`. But shouldn't that already be covered by the `state_advance_timer` which triggers at the same time and calls `fork_choice` which calls `prepare_beacon_proposer_async`?

paulhauner replied:
Good thinking, but unfortunately the `state_advance_timer` only moves forward a single slot:

https://github.com/sigp/lighthouse/blob/c3a793fd73a3b11b130b82032904d39c952869e4/beacon_node/beacon_chain/src/state_advance_timer.rs#L230-L237

So after one skip slot and following slots won't be advanced.

realbigsean replied:
Oh gotcha, didn't realize this

realbigsean commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2229,6 +2229,13 @@ pub fn serve<T: BeaconChainTypes>(
                             )
                         })?;
 
+                    chain.prepare_beacon_proposer_blocking().map_err(|e| {
```
I sort of think this endpoint should only handle updating the proposer preparation cache, and we should make sure fork choice is the only place we call `prepare_beacon_proposer_async` from.  

I could see the benefit of this potentially allowing last second suggested fee recipients changes.  But I'm not sure that's something we should try to code for.  To me it seems like the later this information is provided the less profitable the block could be and the more likely it is to just be missed completely, so there's already strong incentives to provide the fee recipient early and not to change it frequently.

If we are worried about not being able to get the fee recipient on startup for validators proposing immediately, we could also persist this cache or allow the beacon chain to read from the suggested fee recipients file.

realbigsean reviewed (commented):
I had a couple comments around simplifying this, but let me know if there are edge cases I am missing that are informing the design here.

Generally I was thinking the easiest/safest way to maintain BN <> EL consistency is to minimize when we call `notify_forkchoice_updated` and to only call it from `fork_choice`.  We'd be relying on fork choice's ordering which seems like the correct thing to do to me but please correct me here otherwise. 

The one place outside of fork choice where calling `notify_forkchoice_updated` seems unavoidable is when we fallback from `get_payloadV1` due to a payload id missing from our cache ([here](https://github.com/sigp/lighthouse/blob/unstable/beacon_node/execution_layer/src/lib.rs#L435)). But calling a synthetic fork choice update here doesn't really seem like a good thing anyways... could we just run a call to `chain.fork_choice()` in this scenario instead?

If we're able to only call `notify_forkchoice_updated` from fork choice, I think we could then remove all the `notify_forkchoice_updated` specific locking that's added here as well.

paulhauner commented:
> Generally I was thinking the easiest/safest way to maintain BN <> EL consistency is to minimize when we call notify_forkchoice_updated and to only call it from fork_choice. 

Even if we do call it inside fork choice, we still need to deal with ordering. Even if the call to fcU is inside the `fork_choice_internal` function, it still doesn't mean the calls are ordered. Something can call `fork_choice_internal` multiple times, unless we've prevented concurrency internally via some lock. We *could* use the main `chain.fork_choice` lock, however I'm hesitant to hold that lock whilst we call out to remote servers. That is likely to cause long-lived contention.

I'm open to just returning a bool from `prepare_beacon_proposer_blocking` and callers can use that value to call fork choice, if they want. However, I think there's still downsides to that. For example, during `fork_choice_internal`, it seems valuable to me to call fcU as quickly as possible, *then* compute the proposers, *then* maybe call it again. Some ELs might avoid syncing non-head-chains until we provide them a fcU message saying it's the head. In those cases, we want the EL to be aware of the head ASAP so they can start syncing it. In the case where the shuffling isn't in the cache, it seems ideal to send 2x fcU messages; one that alerts about the head change and a later one that informs about the proposer.

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3673,69 +3674,337 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            return Ok(());
+        };
+
+        let finalized_root = if head.finalized_checkpoint.root == Hash256::zero() {
+            // De-alias `0x00..00` to the genesis block root.
+            self.genesis_block_root
+        } else {
+            head.finalized_checkpoint.root
+        };
```
Does this need to be de-aliased? Asking because it ended up not being necessary in the recent optimistic sync PR but am not sure why

paulhauner replied:
Ah yes, good question!

We didn't need to do it in the last PR since we were getting the root from fork choice itself. Fork choice does the de-aliasing for us, so no need there. Here we're getting the root from the `BeaconState` which doesn't do the de-aliasing. I didn't notice this at first and found it in a test!

realbigsean commented on beacon_node/beacon_chain/src/proposer_prep_service.rs:
```diff
@@ -0,0 +1,65 @@
+    let slot_duration = chain.slot_clock.slot_duration();
+
+    loop {
+        match chain.slot_clock.duration_to_next_slot() {
```
If `prepare_beacon_proposer_async` takes 4s to complete, this will wait ~20s to run again, but maybe that's ok?

paulhauner replied:
Good point! In 5ddeae81c I've made it so we spawn the task rather than waiting for it to compete. That should ensure a regular interval in all reasonable scenarios.

realbigsean commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -495,6 +533,62 @@ impl ExecutionLayer {
 
+    /// Register that the given `validator_index` is going to produce a block at `slot`.
+    ///
+    /// The block will be build atop `head_block_root` and the EL will need to prepare an
```
```suggestion
    /// The block will be built atop `head_block_root` and the EL will need to prepare an
```

realbigsean commented:
> Even if we do call it inside fork choice, we still need to deal with ordering.  

Yes I was misunderstanding this. I was looking for a way around the mutex lock on fcU because it'll be called from a lot of different places, but I guess it is the only way to guarantee order.

realbigsean commented:
I think [this call](https://github.com/sigp/lighthouse/blob/unstable/beacon_node/execution_layer/src/lib.rs#L435) also needs to use `update_execution_engine_forkchoice_async` to guarantee ordering

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3673,69 +3674,337 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+        let proposer = if let Some(proposer) = cached_proposer {
+            proposer.index
+        } else {
+            if head_epoch + 2 < prepare_epoch {
```
Will this get hit during sync?

paulhauner replied:
Great point! Resolved in 68e464887.

paulhauner committed:
Spawn proposer prepare routine instead of waiting

paulhauner committed:
Prevent running during sync

paulhauner committed:
Update beacon_node/execution_layer/src/lib.rs

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

paulhauner commented:
> I think [this call](https://github.com/sigp/lighthouse/blob/unstable/beacon_node/execution_layer/src/lib.rs#L435) also needs to use `update_execution_engine_forkchoice_async` to guarantee ordering

I agree, I don't think we can guarantee ordering without adding something there. However, it's complicated because we're producing a payload and using the fcU call only to obtain a payload ID when we don't already have one. Like you mentioned, it's possible that the head changes midway through producing a block and that fcU call is out-of-order. In such a case, you'd expect us to issue the latest head during `get_payload`, however that wouldn't help us get a payload ID for the block we're trying to produce.

So, it seems the choice during `get_payload` is to either:

1. Always send an fork choice update that's potentially out-of-order in order to obtain a payload ID.
1. Only send a fork choice update to get a payload ID *if* it agrees with the current head.

Presently we have (1) and I think that's ideal since we don't risk failing on block production. The downside of using (1) over (2) is that our head might bounce around a little from the EL perspective, which is safe but not ideal.

So, I'm tempted to leave it as-is but I'm definitely keen to hear your thoughts.

paulhauner commented:
Alrighty, I think all comments have been addressed @realbigsean :pray:

realbigsean commented:
> Presently we have (1) and I think that's ideal since we don't risk failing on block production. The downside of using (1) over (2) is that our head might bounce around a little from the EL perspective, which is safe but not ideal.

Yes that's true, I guess in block production it's just most important to build on top of `canonical_head` full stop.

realbigsean committed:
Merge branch 'unstable' into prepare-proposer

realbigsean commented on beacon_node/beacon_chain/tests/payload_invalidation.rs:
```diff
@@ -620,3 +642,55 @@ fn invalid_after_optimistic_sync() {
+            .start_of(next_slot)
+            .unwrap()
+            .as_secs(),
+        random: *head
```
```suggestion
        prev_randao: *head
```

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3653,69 +3659,353 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+                .start_of(prepare_slot)
+                .ok_or(Error::InvalidSlot(prepare_slot))?
+                .as_secs(),
+            random: head.random,
```
```suggestion
            prev_randao: head.random,
```

realbigsean reviewed (commented):
I tried to resolve a merge conflict via github, but there are a couple extra places that need re-naming.

realbigsean committed:
Update beacon_node/beacon_chain/tests/payload_invalidation.rs

realbigsean committed:
Update beacon_node/beacon_chain/src/beacon_chain.rs

realbigsean reviewed (approved):
Ok this looks good to me :slightly_smiling_face:

paulhauner commented:
It turns out there's a bit of an edge-case with the transition block that isn't covered here. I've documented it in #3058.

I think we should merge this PR as-is since it's a significant improvement and has no issues before/after the transition block. We can come back for #3058.

paulhauner committed:
Squash merge of #3057

Squashed commit of the following:

commit c03670e101ac4dfe411c38ce8b450e84528e32e3
Author: Paul Hauner <paul@paulhauner.com>
Date:   Fri Mar 4 09:41:11 2022 +1100

    Filter out zero-hashes

commit 6d269dd77b931d3bb93a7d9bead89933839159bd
Author: Paul Hauner <paul@paulhauner.com>
Date:   Fri Mar 4 09:09:28 2022 +1100

    Avoid sending fcU with zero hash

paulhauner committed:
Add metrics

paulhauner committed:
Merge branch 'unstable' into prepare-proposer

paulhauner commented:
Alrighty, I've been testing this on `merge-devnet-5` with our 1k validators. It's looking good there so I'm happy to merge.

There are two changes since last review:

1. Add in the changes from #3057
2. Add some metrics.

You can see these changes here: https://github.com/paulhauner/lighthouse/compare/76a1d1d...d029c7b

If I can get an approval for these changes then I'll merge :)

paulhauner commented:
This is failing CI due to Geth changing some CLI flags. See https://github.com/sigp/lighthouse/pull/3046#issuecomment-1061292509

paulhauner committed:
Use spawn_blocking for sync fork choice call

paulhauner commented:
I've also added [4811406](https://github.com/sigp/lighthouse/pull/3043/commits/48114062a8f284573b6fe2a9980dd8f1d4048303) after observing a panic:

```
thread 'tokio-runtime-worker' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.'
```

That commit has approval from @AgeManning.

paulhauner committed:
Add payload id hit/miss metrics

paulhauner commented:
Sorry for the noise here. As you can see, I found some bugs whilst testing with our validators on `merge-devnet-5`.

I think this will require additional review, sorry @realbigsean. Unfortunately the diff between the previously-approved changes and the current commit includes a merge of `unstable` (I had to resolve a conflict, sorry).

We should be good to go after this round of review, this branch is performing well presently :rocket:

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into prepare-proposer

 Conflicts:
	Cargo.lock
	beacon_node/execution_layer/Cargo.toml

realbigsean committed:
Cargo.lock

realbigsean commented on beacon_node/beacon_chain/tests/payload_invalidation.rs:
```diff
@@ -85,6 +93,20 @@ impl InvalidPayloadRig {
+            .server
+            .take_previous_request()
+            .expect("no previous request");
+        dbg!(&json);
```
Is this intentionally left or a relic of debugging?
```suggestion
```

realbigsean commented on beacon_node/execution_layer/src/metrics.rs:
```diff
@@ -0,0 +1,34 @@
+    );
+    pub static ref EXECUTION_LAYER_PAYLOAD_ATTRIBUTES_LOOKAHEAD: Result<Histogram> = try_create_histogram(
+        "execution_layer_payload_attributes_lookahead",
+        "Duration between a fcU call with PayloadAttributes and when the block should be produced",
```
```suggestion
        "Duration between an fcU call with PayloadAttributes and when the block should be produced",
```

or `a forkChoiceUpdated`

realbigsean reviewed (approved):
New changes look good to me, left a couple nits.  Resolved a merge conflict in `Cargo.lock` and a `Cargo.toml` as well to try to get CI running

paulhauner committed:
Apply suggestions from code review

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

paulhauner commented:
Thanks for the review and CI fix @realbigsean :pray: 

I'll see how this goes on CI and the testnets and merge if all goes well.

paulhauner commented:
Looks like windows tests are just being dodgy again :(

Everything is looking good on the testnet nodes :tada: 

bors r+
